File: src/Multigraph.cpp
#include "Multigraph.hh"

#include <math.h>
#include <assert.h>
#include <bitset>
#include <complex>
#include <cmath>
#include <queue>

namespace farquaad {


void Multigraph::AddEdge(NodeLabel source, NodeLabel target) {
    assert(source < (NodeLabel) nodesCount and target < (NodeLabel) nodesCount);
    
    if(Multiplicity({source, target}) == 0)
        adjacency[source].emplace_back(target);

    ++edgesCount;
    ++in_degree[target];
    ++out_degree[source];
    ++multiplicity[{source, target}];
}



std::map<Multigraph::Edge, size_t> Multigraph::Eulerify() {
    std::queue<NodeLabel> B_plus, B_minus;

    // Retrieve surplus and deficient nodes
    for(NodeLabel u=0; u<(NodeLabel) NodesCount(); ++u) {
        if(Balance(u) > 0)
            B_plus.emplace(u);
        else if(Balance(u) < 0)
            B_minus.emplace(u);
    }

    std::map<Edge, size_t> E_bottom;

    while(not B_minus.empty()) {
        auto u = B_minus.front();
        B_minus.pop();

        assert(not B_plus.empty());

        auto v = B_plus.front();
        assert(Balance(v) > 0);

        while(Balance(u) < 0) {
            assert(Balance(u) < 0 and Balance(v) > 0);
            
            AddEdge(u, v);
            ++E_bottom[{u, v}];

            if(Balance(v) == 0) {
                B_plus.pop();
                if(not B_plus.empty())
                    v = B_plus.front();
            }
        }
    }

    return E_bottom;
}


} // farquaad




// vector<vector<complex<double> > > Multigraph::getLineGraph() {
//     vector<vector<complex<double> > > lineGraphAdjacency(edgesCount, vector<complex<double> >(edgesCount, complex<double>(0, 0)));

//     vector<vector<int> > rowsEquals(nodesCount, vector<int>(0)); //indexes of all rows of lineGraphAdjacency which are equals because they are all about edge inciding in j.

//     for(int j = 0; j < nodesCount; j++){ // fixed a j as 'middle node', I will handle all equal rows (since all edges will be incident in j)
//                                          // until I change to j+1.
//         for(int i = 0; i < nodesCount; i++){
//             vector<int> edgesFromItoJ = adjacency[i][j]; //edges ENTERING in j from i 

//             for(int k = 0; k < edgesFromItoJ.size(); k++){ // for each edge entering in j from i 
//                 rowsEquals[j].push_back(edgesFromItoJ[k]);
//                 for(int w = 0; w < nodesCount; w++){
//                     vector<int> edgesFromJtoW = adjacency[j][w]; //edges EXITING from j going to w
//                     for(int m = 0; m < edgesFromJtoW.size(); m++){
//                         lineGraphAdjacency[edgesFromItoJ[k]][edgesFromJtoW[m]] = complex<double>(1,0);
//                     }
//                 }
//             }
//         }
//         sort(rowsEquals[j].begin(), rowsEquals[j].end());
//     }

//     for(int i = 0; i < nodesCount; i++){
//         vector<int> sameRowsInducedByI = rowsEquals[i];
//         int size = sameRowsInducedByI.size();
//         int hadamardColumn = 0; 
//         for(int j = 0; j < size; j++){ // for each edge inside sameRowsInducedByI
//             int currentEdge = sameRowsInducedByI[j];
//             for(int k = 0; k < edgesCount; k++){
//                 if(lineGraphAdjacency[currentEdge][k] == complex<double>(1, 0)){
//                     lineGraphAdjacency[currentEdge][k] = dftCoefficient(size, j, hadamardColumn);
//                     hadamardColumn++;
//                 }
//             }
//         }
//     }


//     return lineGraphAdjacency;
// }

File: Python - src/Multigraph.py
from Edge import Edge
import numpy as np

class Multigraph: 

    def __init__(self):
        self.edges = 0
        self.adj_list = { }
        self.multiplicity = { }
        self.in_degree = { }
        self.out_degree = { }


    def indegree(self, n):
        return self.in_degree.get(n, 0)
    
    def outdegree(self, n):
        return self.out_degree.get(n, 0)

    def balance(self, n):
        return self.outdegree(n) - self.indegree(n)

    def edge_multiplicity(self, e):
        return self.multiplicity.get(e, 0)

    def edge_count(self):
        return self.edges

    def init_node(self, node):
        if not node in self.adj_list.keys():
            self.adj_list[node] = []
    
        if not node in self.in_degree.keys():
            self.in_degree[node] = 0
        
        if not node in self.out_degree.keys():
            self.out_degree[node] = 0
        

    def add_edge(self, source, target):

        self.init_node(source)
        self.init_node(target)

        if not Edge(source, target) in self.multiplicity.keys():
            self.multiplicity[Edge(source, target)] = 0 

        if target not in self.adj_list[source]:
            self.adj_list[source].append(target)

        self.edges += 1
        self.in_degree[target] += 1 
        self.out_degree[source] += 1 
        self.multiplicity[Edge(source, target)] += 1


    def pretty_print(self):
        for key in self.adj_list.keys():
            print(str(key) + ": " + str(self.adj_list[key]))

    def outgoing_edges(self, u):
        res = []
        for v in self.adj_list[u]:
            e = Edge(u, v)
            res += [(e, k) for k in range(self.edge_multiplicity(e))]
        return res

    def all_edges(self):
        res = []
        for u in self.adj_list.keys():
            res += self.outgoing_edges(u)

        return res

    def eulerify(self):
        B_plus = [u for u in self.adj_list.keys() if self.balance(u) > 0]
        B_minus = [u for u in self.adj_list.keys() if self.balance(u) < 0]
        E_bottom = []
        first = 0

        for u in B_minus:
            while self.balance(u) < 0:
                v = B_plus[first]
                self.add_edge(u, v)
                E_bottom.append((u, v))

                if self.balance(v) == 0:
                    first += 1

        return E_bottom

    def to_adjacency_matrix(self):
        adj_m = np.array([[self.edge_multiplicity(Edge(u, v)) for v in self.adj_list.keys()] for u in self.adj_list.keys()])

        return adj_m

    def line_graph(self):
        line = Multigraph()

        edges = self.all_edges()

        for i in edges:
            v = i[0].target
            for k in self.outgoing_edges(v):
                line.add_edge(i, k)

        return line